src/v/cluster/jumbo_log_stm.cc

```diff
@@ -0,0 +1,221 @@
+    ::jumbo_log::write_intent_segment segment;
+};
+
+using cmd_key = named_type<uint8_t, struct cmd_key_tag>;
+struct jumbo_log_stm::create_write_intent_cmd {
+    static constexpr cmd_key key{0};
+    using value = jumbo_log_stm::write_intent;
+};
+
+jumbo_log_stm::jumbo_log_stm(ss::logger& logger, raft::consensus* c)
+  : jumbo_log_stm(logger, c, config::shard_local_cfg()) {}
+
+jumbo_log_stm::jumbo_log_stm(
+  ss::logger& logger, raft::consensus* c, config::configuration& cfg)
+  : persisted_stm(jumbo_log_snapshot, logger, c)
+  , _next_write_intent_id(1) {
+    std::ignore = cfg;
+}
+
+// TODO(nv): Fix all raft::make_error_code calls to use proper error codes.
+//           Also, consider using std::error_code instead of std::system_error.
+
+ss::future<jumbo_log::write_intent_id_t> jumbo_log_stm::create_write_intent(
+  jumbo_log::segment_object object, model::timeout_clock::duration timeout) {
+    // TODO(nv): this lock needs a timeout
+    auto u = co_await _lock.get_units();
+
+    if (!co_await sync(timeout)) {
+        throw std::system_error(raft::make_error_code(raft::errc::timeout));
+    }
+
+    // If this is a retry, just return the existing write intent id.
+    auto write_intent_id_it = _write_intent_id_by_object_id.find(object.id);
+    if (write_intent_id_it != _write_intent_id_by_object_id.end()) {
+        // Ensure that size_bytes also matches and the client is not reusing
+        // object ids.
+        if (auto write_intent_it = _write_intents.find(
+              write_intent_id_it->second);
+            write_intent_it != _write_intents.end()
+            && write_intent_it->second.object.size_bytes != object.size_bytes) {
+            throw std::system_error(
+              raft::make_error_code(raft::errc::invalid_input_records));
+        }
+        co_return write_intent_id_it->second;
+    }
+
+    storage::record_batch_builder batch_builder(
+      model::record_batch_type::jumbo_log_cmd, model::offset{});
+
+    auto intent_id = _next_write_intent_id;
+
+    auto record = create_write_intent_cmd::value{
+      .segment = jumbo_log::write_intent_segment(intent_id, object)};
+
+    batch_builder.add_raw_kv(
+      serde::to_iobuf(create_write_intent_cmd::key), serde::to_iobuf(record));
+
+    // No timeouts here! We can't allow concurrency as it will break write
+    // intent id allocation invariants.
+    auto r = co_await _raft->replicate(
+      _insync_term,
+      model::make_memory_record_batch_reader(std::move(batch_builder).build()),
+      raft::replicate_options(raft::consistency_level::quorum_ack));
+    if (!r) {
+        throw std::system_error(raft::make_error_code(raft::errc::timeout));
+    }
+
+    // TODO(nv): Sure ok to timeout here?
+    if (!co_await wait_no_throw(
+          model::offset(r.value().last_offset()),
+          model::timeout_clock::now() + timeout)) {
+        throw std::system_error(raft::make_error_code(raft::errc::timeout));
+    }
+
+    co_return intent_id;
```

Comment:
my first impression here is that you can get rid of the locking in this method by restructing the way this works:

1. assign write intent ids on log replay, instead of when you replicate data into the log.
2. so you replicate the create-write-intent-command
3. sync / wait for replay
4. look-up the result of the command being applied to the stm.
5. no locking necessary because there is only a single writer to the shared state, which is the stm replay logic.